packages/hint-compat-api/src/css-base.ts

```diff
@@ -0,0 +1,160 @@
+import { HintContext } from 'hint/dist/src/lib/hint-context';
+import { IHint } from 'hint/dist/src/lib/types';
+import { StyleParse, StyleEvents } from '@hint/parser-css/dist/src/types';
+import { CompatApi, CompatCSS } from './helpers';
+import { FeatureInfo, BrowsersInfo, SupportStatementResult } from './types';
+import { SimpleSupportStatement, VersionValue, SupportBlock, SupportStatement } from './types-mdn.temp';
+
+import { browserVersions } from './helpers/normalize-version';
+import { CSSFeatureStatus } from './enums';
+
+export default abstract class BaseCompatApiCSS implements IHint {
+    private readonly statusName: CSSFeatureStatus;
+    private compatApi: CompatApi;
+    private compatCSS: CompatCSS;
+
+    abstract getFeatureVersionValueToAnalyze(browserFeatureSupported: SimpleSupportStatement): VersionValue;
+    abstract isSupportedVersion(browser: BrowsersInfo, feature: FeatureInfo, currentVersion: number, version: number): boolean;
+    abstract isVersionValueSupported(version: VersionValue): boolean;
+    abstract isVersionValueTestable(version: VersionValue): boolean;
+
+    public constructor(context: HintContext<StyleEvents>, statusName: CSSFeatureStatus, isCheckingNotBroadlySupported: boolean) {
+        this.compatApi = new CompatApi('css', context, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, this.testFeatureIsSupported.bind(this));
+        this.statusName = statusName;
+
+        context.on('parse::end::css', this.onParseCSS.bind(this));
+    }
+
+    private async onParseCSS(styleParse: StyleParse): Promise<void> {
+        const { resource } = styleParse;
+
+        this.compatCSS.setResource(resource);
+        await this.compatCSS.searchCSSFeatures(this.compatApi.compatDataApi, styleParse);
+    }
+
+    private async testFeatureIsSupported(feature: FeatureInfo, supportBlock: SupportBlock): Promise<void> {
+        const browsersToSupport = Object.entries(supportBlock).filter(([browserName, _]: [string, SupportStatement]): boolean => {
+            return this.compatApi.isBrowserIncludedInCollection(browserName);
+        });
+
+        const groupedSupportByBrowser = browsersToSupport.reduce((group, browserInfo, _, a1) => {
+            return this.groupSupportStatementByBrowser(feature, group, browserInfo);
+        }, {});
+
+        const supportStatementResult: SupportStatementResult = {
+            browsersToSupportCount: browsersToSupport.length,
+            groupedBrowserSupport: groupedSupportByBrowser,
+            notSupportedBrowsersCount: Object.keys(groupedSupportByBrowser).length
+        };
+
+        const hasImcompatibleBrowsers: boolean = supportStatementResult.notSupportedBrowsersCount > 0;
+
+        if (!hasImcompatibleBrowsers) {
+            return;
+        }
+
+        const message: string = this.generateReportErrorMessage(feature, supportStatementResult);
```

Comment:
again, string could be removable (probably)

Reply:
👍